Tests/GRPCTests/MutualTLSTests.swift

```diff
@@ -0,0 +1,156 @@
+
+    let server = try! Server.start(configuration: serverConfiguration).wait()
+    let port = server.channel.localAddress!.port!
+
+    // Setup the client.
+    var clientConfiguration = ClientConnection.Configuration.default(
+      target: .hostAndPort("localhost", port),
+      eventLoopGroup: clientEventLoopGroup
+    )
+    clientConfiguration.tlsConfiguration = clientTLSConfiguration
+    clientConfiguration.connectionBackoff = nil
+    clientConfiguration.backgroundActivityLogger = self.clientLogger
+    let clientErrorExpectation = self.expectation(description: "client error")
+    clientErrorExpectation.isInverted = expectedOutcome == .success
+    clientErrorExpectation.assertForOverFulfill = false
+    let clientErrorDelegate = ErrorRecordingDelegate(expectation: clientErrorExpectation)
+    clientConfiguration.errorDelegate = clientErrorDelegate
+
+    let client = Echo_EchoClient(channel: ClientConnection(configuration: clientConfiguration))
+
+    // Make the call.
+    let call = client.get(.with { $0.text = "mumble" })
+
+    // Wait for side effects.
+    self.wait(for: [clientErrorExpectation, serverErrorExpectation], timeout: 1)
+
+    switch expectedOutcome {
+    case .success:
+      // Verify no errors.
+      XCTAssert(serverErrorDelegate.errors.isEmpty)
+      XCTAssert(clientErrorDelegate.errors.isEmpty)
+      // Verify response.
+      let response = try call.response.wait()
+      XCTAssertEqual(response.text, "Swift echo get: mumble")
+      let status = try call.status.wait()
+      XCTAssertEqual(status.code, .ok)
+    case .serverError:
+      // Verify handshake error.
+      guard case .handshakeFailed = serverErrorDelegate.errors.first as? NIOSSLError else {
+        XCTFail("Expected NIOSSLError.handshakeFailed")
+        return
+      }
+    case .clientError:
+      // Verify handshake error.
+      guard case .handshakeFailed = clientErrorDelegate.errors.first as? NIOSSLError else {
+        XCTFail("Expected NIOSSLError.handshakeFailed")
+        return
+      }
+    }
+  }
+
+  func test_ValidClientAndServerCerts_Success() throws {
+    let serverTLSConfiguration = GRPCTLSConfiguration.makeServerConfigurationBackedByNIOSSL(
+      certificateChain: [.certificate(SampleCertificate.server.certificate)],
+      privateKey: .privateKey(SamplePrivateKey.server),
+      trustRoots: .certificates([SampleCertificate.ca.certificate]),
+      certificateVerification: .noHostnameVerification
+    )
+    let clientTLSConfiguration = GRPCTLSConfiguration.makeClientConfigurationBackedByNIOSSL(
+      certificateChain: [.certificate(SampleCertificate.client.certificate)],
+      privateKey: .privateKey(SamplePrivateKey.client),
+      trustRoots: .certificates([SampleCertificate.ca.certificate]),
+      certificateVerification: .fullVerification
+    )
+    try self.performTestWith(serverTLSConfiguration, clientTLSConfiguration, expect: .success)
+  }
+
+  func test_noServerCert_ClientError() throws {
```

Comment:
I'm not sure how useful this test is as an _mTLS test_: it's validates that a client configured with TLS fails to connect to a plaintext server.